source/birchwood/protocol/formatting.d

```diff
@@ -1,2 +1,99 @@
 module birchwood.protocol.formatting;
 
+import std.string;
+
+// Reset character; resets all formatting
+enum reset_code = '\x0F';
+
+// Toggle characters
+enum bold_code = '\x02';
+enum italic_code = '\x1D';
+enum underline_code = '\x1F';
+enum strikethrough_code = '\x1E';
+enum monospace_code = '\x11';
+enum reverse_colors_code = '\x16'; // NOT UNIVERSALLY SUPPORTED
+
+// Color characters
+enum ascii_color_code = '\x03';
+enum hex_color_code = '\x04';
+
+// Simple color codes
+enum simple_colors: string {
+    WHITE = "00",
+    BLACK = "01",
+    BLUE = "02",
+    GREEN = "03",
+    RED = "04",
+    BROWN = "05",
+    MAGENTA = "06",
+    ORANGE = "07",
+    YELLOW = "08",
+    LIGHT_GREEN = "09",
+    CYAN = "10",
+    LIGHT_CYAN = "11",
+    LIGHT_BLUE = "12",
+    PINK = "13",
+    GREY = "14",
+    LIGHT_GREY = "15",
+    DEFAULT = "99" // NOT UNIVERSALLY SUPPORTED
+}
+
+// Return the hex control character if color is a hexadecimal color code, the ASCII control character if color is two ASCII digits, and throw an exception if it's neither
+// This function might be useless now that set_fg and set_fg_bg have been changed, but I'll keep it in case it's needed later.
+char generate_color_control_char(string color) {
+    if (color.length == 6) {
+        return hex_color_code;
+    } else if (color.length == 2) {
+        return ascii_color_code;
+    } else {
+        throw new StringException("Invalid color code (must be either two ASCII digits or a hexadecimal code of the form RRGGBB)");
+    }
+}
+
+// Generates a string that changes the foreground color
+string set_foreground(string color) {
```

Comment:
Maybe the function should take in the enum rather (`simple_colors`) (I would rename it to `SimpleColor`) and then it should convert to a string (seeing as the members are just strings replaced at compilation time).

Reply:
The reason I say this is because using the enum is neater and also guarantees compilation type-safety in that only members of valid colors can be passed in **and** because of this helps with IDE auto-completion/suggestions.

Reply:
> I wrote the function to take in a string so it can also take in hexadecimal color codes (RRGGBB). I could make separate functions for colors formatted as hexadecimal codes and ASCII digit pairs.

Could you add a wrapper function that does as described please. Do keep the current ones however, as the custom RRGGBB is useful feature to have.

Reply:
I've added a function that does as described. It isn't a wrapper due to the current ones' if statements.